src/enumerable.cr

```diff
@@ -2292,7 +2292,11 @@ module Enumerable(T)
     # if the type is a union.
     def self.first
       {% if X.union? %}
-        {{X.union_types.first}}
+        {{ 
+          raise("Enumerable#sum() does support Union types. Instead, " +
+                 "use Enumerable#sum(initial) with an initial value of " +
+                 "the expected type of the sum call.") 
```

Comment:
**issue:** `Reflect#first` is used for `Enumerable#product` as well as `#sum` (this is mentioned in the issue). So the error message should reflect that. And we should have a spec for `#product`.

Reply:
Good catch!  Done :)